### DISCUSSION TITLE
OpenTofu / Terraform provider planned to be officially supported?

### DISCUSSION THREAD

**Discussion by dragetd** [openbao/openbao#637] [General]
A big strength of a secret management engine beyond simple password managers is the IaC approach of managing them. So generating infrastructure and putting the secrets into the secret store via tofu for example.

Vault had a provider that would allow this. I am pretty sure *some* OpenBAO provider will eventually pop up, but are there any plans to make it officially part of the project? (With implications like some time in the future releases waiting on each other to ensure compatibility or even inclusion in tests etc.)?

**Comment by cipherboy** [comment]
\o hey @dragetd -- we had discussed this a bit in https://github.com/openbao/openbao/issues/339 but I think the general points still stand: until upstream is unwilling to merge net-new features to the provider and/or OpenBao diverges too far in new functionality, my gut says we should probably keep encouraging collaboration with them on some of this common infra & tooling. 

That said, if that doesn't occur or we find we've added enough new functionality that we think we could take advantage of, I'm happy to entertain blessing a fork. 

What's your thoughts?

**Comment by dragetd** [comment]
Oh, actually it did not occur to me that the vault provider might just work, as the API is still in place!
Silly me. So I guess I can continue kind of 'mostly anything right now' and just throw it against OpenBAO instead of Vault.

Yes, unless the feature-set diverges noticeably, there is not really a point in investing that work right now. Focus should be on polishing it up as it's own project.

I'll see what happens if I spin up openbao instead of my vault and throw my tofu against it. More food for my infrastructure. xD

**Reply by cipherboy** [reply]
Enough to make one hungry! :D good luck, and let us know if you encounter anything!

**Comment by ttiborgh** [comment]
Hi @dragetd, @cipherboy! We plan to install OpenBao into our environment with the Vault provider of Terraform. We are curious if you guys have experienced any strange behaviour when installing OpenBao with Terraform? Are there any issues you are aware of? We would like to make sure to be prepared for any issues or blockers in advance. :)

**Comment by andreaso** [comment]
@ttiborgh: I'm currently using the Terraform Vault provider together with a small/simple OpenBao instance, and so far I haven't had any issues.

**Comment by LordGaav** [comment]
One thing that might be an issue: the new CEL functionality in the PKI engine doesn't seem to be reachable using the Terraform provider.

**Comment by jficz** [comment]
We're encountering bugs in the upstream provider quite often which sit there for years without much interest from Hashi and we have some feature requests _with PRs_ which are completely [ignored](https://github.com/hashicorp/terraform-provider-vault/issues/2541). We're actually considering a soft internal fork with the bug fixes and features implemented as [some](https://github.com/hashicorp/terraform-provider-vault/issues/1885) are quite serious.

We'd be more than happy to submit our PRs to a fork which actually works with their community.

**Reply by cipherboy** [reply]
@jficz Do you have thoughts on fork maintenance?

I'm kinda wondering if we want a new format, repo with patches and a renamer utility, that auto-generates the destination repo.

Are there patterns for lightweight forks for source repos I wonder? Especially when we want to follow upstream? 🤔

**Reply by jficz** [reply]
Not really, it heavily depends on what the fork's goal would be imho. In our internal case it's just about a few relatively simple fixes for a few annoying things so we would probably go with rolling rebase.

But if the fork was to be an "Official Bao Provider" then it would be more complex - as I [understand](https://github.com/openbao/openbao/issues/339) there are a few different paths to choose from:

1. entirely new provider based on some kind of automated api generator
2. a "wrapper" on top of upstream provider
3. a soft / hard fork

I agree with a couple of comments in the issue that option 1 would probably be a bad idea for various reasons - mostly migration from current setups and the fact that the tooling appears not to be maintained.

The wrapper approach would be interesting but then I think a line would have to be drawn between what would be wrapped and what not. The obvious would be to just add bao-specific stuff and keep the upstream provider for everything else. That would mean we probably could not do things like fixing bugs that upstream ignores or adding features to existing resources which are not bao-specific (basically the two issues I linked above) but it would also mean easier maintenance in the long term with higher initial costs.

A soft fork seems like the easiest way to start at least - we wouldn't have to reinvent anything, just patch in/out things the community wants to. Lower initial costs with the risk of higher maintenance overhead in the future.

Both wrapper and soft forks are ultimately bets on what happens in the future both with the community requirements and upstream source. And I'm not sure I'm qualified comment on a betting strategy :)

**Comment by manzsolutions-lpr** [comment]
The latest release introduces new security options which are enabled by default but cannot be disabled using Hashicorp’s Vault Terraform provider since it does not support the options, i.e.:

- allow_wildcards_in_identity_templates
- allow_slashes_in_identity_templates
- allow_globs_in_identity_templates
- allow_commas_in_identity_templates

This may effectively brick the provider for existing implementations.

https://github.com/openbao/openbao/releases/tag/v2.6.0

**Reply by satoqz** [reply]
> That said, if that doesn't occur or we find we've added enough new functionality that we think we could take advantage of, I'm happy to entertain blessing a fork.

@cipherboy it might be time by now, either way. We also have additional namespace APIs, the workflow APIs and upcoming external key APIs to cover, just to name some.

**Comment by mcmdev** [comment]
Setting the `callbackmode` for the JWT/OIDC auth method also isn't possible using the Vault terraform provider.

**Reply by jficz** [reply]
> Setting the `callbackmode` for the JWT/OIDC auth method also isn't possible using the Vault terraform provider.

There are bound to be more, Bao has already diverged from upstream quite a lot. Time to make a decision imho.

**Reply by cipherboy** [reply]
I think from my side what remains is interest in contributing and how to manage the fork; I seem to recall there were thoughts on making synchronization easier. If someone wants to put forward a proposal, that'd be great.

**Reply by jficz** [reply]
> If someone wants to put forward a proposal, that'd be great.
 What would you like to have in the proposal?

**Reply by cipherboy** [reply]
@jficz @satoqz I'd like to see the following:

- Discussion of implementation approaches. So far we've talked about (between here and #339): a fully greenfield provider (+/- being auto-generated from OpenAPI); a mini-provider which just implements net-new objects; a hard fork of upstream TF provider; and probably others I'm forgetting about. Need to pick one.
- What migration from upstream would look like, whatever that means. TF import? Module replacement? Something else?
- Discussion of how we take patches if we're going the hard-fork route. Is it a pointer + patches which gets auto-generated into a full repo? Do we use merge commits? Rebase? Cherry-pick relevant patches? How much of this can we automate? &c.  ---> Looking in general for an approach that will help us scale across g-k-w and openbao-helm to name a few.
- Discussion of how OpenAPI specs are improved/validated/... if we're going that route. From my PoV we still need a lot of love there.
- Would we support just OpenBao or both?  How would we maintain/test compatibility in the latter case? 

If someone wants to make a recommendation, that'd be great.

**Reply by karras** [reply]
See https://github.com/openbao/openbao/pull/3720 for @Luiggi33's RFC PR.
